src/mapping/query/function/mod.rs

```diff
@@ -120,6 +122,70 @@ build_signatures! {
     contains => ContainsFn,
     slice => SliceFn,
     tokenize => TokenizeFn,
+    split => SplitFn,
+}
+
+pub(in crate::mapping) struct TypedArgument {
+    pub resolver: Box<dyn Function>,
+    pub parameter: Parameter,
+}
+
+impl TypedArgument {
+    pub fn new(resolver: Box<dyn Function>, parameter: Parameter) -> Self {
+        Self {
+            resolver,
+            parameter,
+        }
+    }
+}
+
+// delegates to resolver to satisfy tests in `mapping::parser`.
+impl std::fmt::Debug for TypedArgument {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.resolver.fmt(f)
+    }
+}
+
+impl Function for TypedArgument {
+    fn execute(&self, ctx: &Event) -> Result<Value> {
+        let value = self.resolver.execute(ctx)?;
+
+        // Ask the parameter if it accepts the given value.
+        if !(self.parameter.accepts)(&value) {
+            return Err(format!(
+                "invalid argument type '{}' for parameter '{}'",
+                value.kind(),
+                self.parameter.keyword
+            ));
+        }
+
+        Ok(value)
+    }
+}
+
+/// Because the Regex object doesn't contain a global flag
+/// (this is determined by calling either replace or replace_all)
+/// we need to wrap the object and store this flag when we parse it.
+#[derive(Debug, Clone)]
+pub(in crate::mapping) struct RemapRegex {
+    pub regex: regex::Regex,
+    pub global: bool,
+}
+
+#[derive(Debug)]
+pub(in crate::mapping) enum ArgumentKind {
+    Value(Box<dyn Function>),
+    Regex(RemapRegex),
+}
+
+impl ArgumentKind {
+    pub fn into_value(self) -> Result<Box<dyn Function>> {
+        if let ArgumentKind::Value(value) = self {
+            Ok(value)
+        } else {
+            Err("expected value".into())
+        }
```

Comment:
Match all things :grin: 
```suggestion
        match self {
            ArgumentKind::Value(value) => Ok(value),
            _ => Err("expected value".into()),
        }
```

Reply:
I so much prefer the first version, using `if let` when just matching against a single variant.

What is the reason you would prefer to match here?

Reply:
I have come to prefer `match` over `if let … else`. It seems more idiomatic (matching the state of an enum), easier to add cases later, and takes fewer lines to boot in cases like this.

Reply:
Making it easier to add cases later is a very good point. I'm sold!